src/mpi/init/init_thread_cs.c

```diff
@@ -28,25 +54,15 @@ void MPII_init_thread_and_enter_cs(int thread_required)
     MPID_Thread_mutex_create(&MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX, &err);
     MPIR_Assert(err == 0);
 
-    MPIR_ThreadInfo.isThreaded = (thread_required == MPI_THREAD_MULTIPLE);
-    required_is_threaded = MPIR_ThreadInfo.isThreaded;
+    MPIR_THREAD_CS_ENTER(MPIR_THREAD_GLOBAL_ALLFUNC_MUTEX);
```

Comment:
Overall, does it solve the problem? I think the fundamental issue is what this mutex intends to protect.
This unconditional `CS_ENTER` and `CS_EXIT` can only protect `MPII_init` and `MPII_finalize` functions since other attempts to take this mutex (at other places) will check `MPIR_ThreadInfo.isThreaded` and they ignore this mutex because it is temporarily set to 0.

If this mutex only needs to protect `MPII_init` and `MPII_finalize`, this fix should work.

Reply:
It only protects `MPI_Init/Finalize` from itself. Even if we remove the lock, it is still correct per MPI standards -- attributing to user errors. But as implementation, we try to be nicer when we can.